File: p1.py
#Time: O(V+E)
#Space: O(V+E)
class Solution:

    def __init__(self):
        self.time = 0

    def getGraph(self, n: int, connections: List[List[int]]) -> List[List[int]]:

        #   a helper function to create the graph
        adjacencyList = []
        for i in range(n):
            adjacencyList.append([])

        for edge in connections:
            u = edge[0]
            v = edge[1]
            adjacencyList[u].append(v)
            adjacencyList[v].append(u)

        return adjacencyList

    def dfsVisit(self, adjacencyList: List[List[int]], u: int, visited: List[bool],
                   discoveryTime: List[int], lowTime: List[int], parent: List[int],
                   finalResult: List[List[int]]) -> None:

        
        visited[u] = True
        self.time += 1
        discoveryTime[u] = self.time
        lowTime[u] = self.time

        #   for each of its vertices
        for v in adjacencyList[u]:

            #   if not visited
            if (visited[v] != True):
                parent[v] = u                                          

                self.dfsVisit(adjacencyList, v, visited, discoveryTime,
                                lowTime, parent, finalResult)

               
                lowTime[u] = min(lowTime[u], lowTime[v])

                
                if (lowTime[v] > discoveryTime[u]):
                    finalResult.append([u, v])

            
            elif (v != parent[u]):
                lowTime[u] = min(lowTime[u], discoveryTime[v])

    def criticalConnections(self, n: int, connections: List[List[int]]) -> List[List[int]]:

       
        if (connections == None or len(connections) == 0 or n == 0):
            return []

        
        finalResult = []

        adjacencyList = self.getGraph(n, connections)

        visited = [False for i in range(n)]
        discoveryTime = [0 for i in range(n)]
        lowTime = [0 for i in range(n)]
        parent = [0 for i in range(n)]

        
        self.dfsVisit(adjacencyList, 0, visited, discoveryTime,
                        lowTime, parent, finalResult)

        
        return finalResult

File: p1.java
// Time Complexity :O(n) where n is the number of connectios
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no


// Your code here along with comments explaining your approach

class Solution {
    List<List<Integer>> result;
    HashMap<Integer, List<Integer>> map;
    int[] distance;
    int[] lowest;
    int time = 0;
    public List<List<Integer>> criticalConnections(int n, List<List<Integer>> connections) {
        result = new ArrayList<>();
        map = new HashMap<>();
        distance = new int[n];
        //Make the distance array as -1 which will act as not visited
        for(int i=0; i< n; i++){
            distance[i] = -1;
        }
        lowest = new int[n];

        //Make the indegrees hashmap
        for(List<Integer> c: connections){
            List<Integer> temp = map.getOrDefault(c.get(0), new ArrayList<>());
            temp.add(c.get(1));
            map.put(c.get(0), temp);

            temp = map.getOrDefault(c.get(1), new ArrayList<>());
            temp.add(c.get(0));
            map.put(c.get(1), temp);
        }
        //Call recursive function
        dfs(0,0);
        return result;
    }

    private void dfs(int v, int u){
        //base
        if(distance[v] != -1){
            return;
        }
        distance[v] = time;
        lowest[v] = time;
        time++;
        List<Integer> temp = map.get(v);
        //Go through every connection for a node
        for(int i: temp){
            //If it's the parent, do nothing
            if(i == u){
                continue;
            } 
            dfs(i, v);
            //If we find a node which is not a part of any cycle, add it to the result
            if(lowest[i] > distance[v]){
                List<Integer> r = new ArrayList<>();
                r.add(v);
                r.add(i);
                result.add(r);
            }
            //If we detect cycle, mark the lowest
            if(lowest[i] < lowest[v]){
                lowest[v] = lowest[i];
            }
        }
        return;
    }
}